sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -1043,25 +1044,35 @@ public void publishMessage(T input) throws JMSException, JmsIOException {
         }
       }
 
-      public void close() throws JMSException {
-        isProducerNeedsToBeCreated = true;
+      void startProducer() throws JMSException {
+        this.producer = this.session.createProducer(null);
+      }
+
+      void closeProducer() throws JMSException {
         if (producer != null) {
           producer.close();
           producer = null;
         }
-        if (session != null) {
-          session.close();
-          session = null;
-        }
-        if (connection != null) {
-          try {
-            // If the connection failed, stopping the connection will throw a JMSException
-            connection.stop();
-          } catch (JMSException exception) {
-            LOG.warn("The connection couldn't be closed", exception);
+      }
+
+      void close() {
+        try {
+          if (producer != null) {
```

Comment:
may call closeProducer() here. Making producer is opened / closed by single code path